test/CoinSelection/RoundRobin.purs

```diff
@@ -0,0 +1,211 @@
+          runRoundRobin state ((const Nothing) <$ processors) === state
+    test "iterationCount"
+      $ quickCheck
+      $ \(initialState :: SimpleMockRoundRobinState) -> (===)
+          ( length $
+              (unwrap $ runMockRoundRobin initialState).accumulatedEntries
+          )
+          (sum (unwrap initialState).processorLifetimes)
+    test "iterationOrder" $
+      quickCheck prop_runRoundRobin_iterationOrder
+    test "generationCount" $
+      quickCheck prop_runRoundRobin_iterationOrder
+    test "generationOrder" $
+      quickCheck prop_runRoundRobin_generationOrder
+
+-- Tests
+
+prop_runRoundRobin_iterationOrder :: SimpleMockRoundRobinState -> Result
+prop_runRoundRobin_iterationOrder initialState =
+  sortDescending entries === entries
+  where
+  entries = swap <$>
+    (unwrap $ runMockRoundRobin initialState).accumulatedEntries
+  sortDescending = sortBy \x y -> invert $ compare x y
+
+prop_runRoundRobin_generationCount :: SimpleMockRoundRobinState -> Result
+prop_runRoundRobin_generationCount initialState =
+  Map.filter (_ > 0) (unwrap initialState).processorLifetimes
+    === generationCounts
+  where
+  finalState = runMockRoundRobin initialState
+
+  generationCounts :: Map.Map TokenName Int
+  generationCounts = (unwrap finalState).accumulatedEntries
+    # groupByKey
+    # map length
+
+prop_runRoundRobin_generationOrder :: SimpleMockRoundRobinState -> Result
+prop_runRoundRobin_generationOrder initialState =
+  result <?> "initialState: " <> show initialState
+  where
+  finalState = runMockRoundRobin initialState
+
+  generations :: Map.Map Int (Set.Set TokenName)
+  generations = (unwrap finalState).accumulatedEntries
+    # map swap
+    # groupByKey
+    # map Set.fromFoldable
+  result = all (uncurry (flip Set.subset))
+    $ consecutivePairs
+    $ snd <$> Map.toUnfoldable generations
+
+-- Utilites for tests
+-- TODO: where to place them
+
+runRoundRobin
+  :: forall (s :: Type)
+   . s
+  -> Array (s -> Maybe s)
+  -> s
+runRoundRobin state processors =
+  unwrap $ runRoundRobinM state $ (map Identity) <$> processors
+
+consecutivePairs :: forall a. List a -> List (a /\ a)
+consecutivePairs xs = case tail xs of
+  Nothing -> fromFoldable []
+  Just ys -> xs `zip` ys
+
+groupByKey :: forall k v. Ord k => List (k /\ v) -> Map.Map k (List v)
+groupByKey = foldl acc Map.empty
+  where
+  acc :: Map.Map k (List v) -> (k /\ v) -> Map.Map k (List v)
+  acc m (k /\ v) = Map.alter (Just <<< (maybe (singleton v) ((:) v))) k m
+
+ungroupByKey :: forall k v. Map.Map k (List v) -> List (k /\ v)
+ungroupByKey m = do
+  (k /\ vs) <- Map.toUnfoldable m
+  v <- vs
+  pure (k /\ v)
+
+-- Mock state used for testing
+
+-- From https://github.com/input-output-hk/cardano-wallet/blob/9d73b57e23392e25148cfc8db560cb8f656cb56a/lib/coin-selection/test/spec/Cardano/CoinSelection/BalanceSpec.hs#L4036
+newtype MockRoundRobinState k n = MockRoundRobinState
+  { processorLifetimes :: Map.Map k n
```

Comment:
Unnecessary qualified import of `Map` type constructor

Reply:
But I do not see if it clutters code anyhow.